litgpt/data/text_files.py

```diff
@@ -0,0 +1,159 @@
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
+import glob
+import os
+from dataclasses import dataclass, field
+from functools import partial
+from pathlib import Path
+from tqdm import tqdm
+from typing import Optional
+
+from torch.utils.data import DataLoader
+
+from litgpt import Tokenizer
+from litgpt.data import DataModule
+
+
+def get_half_workers():
+    num_workers = (os.cpu_count() - 1) // 2
+    if num_workers < 1:
+        return 1
+    else:
+        return num_workers
+
+
+@dataclass
+class TextFiles(DataModule):
+    """The TextFile data module used for pretraining.
+
+    Reads in text data from plaintext files contained in a data folder
+    and provides training and validation dataloaders that return batches of tokens.
+    Every sample is set to a fixed length.
+    """
+    train_data_path: Path
+    """The path to the data directory used for training that contains .txt files"""
+    val_data_path: Optional[str] = None
+    """The path to the data directory used for validation that
+    contains .txt files. Splits off data for validation from the
+    training set if None."""
+    seed: int = 42
+    """The seed to use for shuffling the dataset."""
+    num_workers: Optional[int] = None
+    """The number of workers to use for data loading.
+       Sets the number of workers equal to the number of avaialable CPUs-1 by default."""
```

Comment:
I strongly advise against this. Say we are training with 8 GPUs in a machine, then in total we would have `(CPUs - 1) * 8` just for dataloading workers which is greater than the number of CPUs available on the system and that's generally not ideal.

As for the preprocessing it's a different thing, because we only have one main process from which we create workers. So there, let's use all CPU cores (see tinystories.py)?